source/extensions/common/wasm/wasm.cc

```diff
@@ -1571,50 +1565,62 @@ bool Wasm::initialize(const std::string& code, bool allow_precompiled) {
-    start = decodeVarint(start, end, &emscripten_metadata_major_version_);
-    start = decodeVarint(start, end, &emscripten_metadata_minor_version_);
-    start = decodeVarint(start, end, &emscripten_abi_major_version_);
-    start = decodeVarint(start, end, &emscripten_abi_minor_version_);
-    uint32_t temp;
-    if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 1) {
-      // metadata 0.2 - added: wasm_backend.
-      start = decodeVarint(start, end, &temp);
+    auto ok = wasm_vm_->load(code, allow_precompiled);
+    if (!ok) {
+      return false;
     }
-    start = decodeVarint(start, end, &temp);
-    start = decodeVarint(start, end, &temp);
-    if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 0) {
-      // metadata 0.1 - added: global_base, dynamic_base, dynamictop_ptr and tempdouble_ptr.
-      start = decodeVarint(start, end, &temp);
+    auto metadata = wasm_vm_->getCustomSection("emscripten_metadata");
+    if (!metadata.empty()) {
+      // See https://github.com/emscripten-core/emscripten/blob/incoming/tools/shared.py#L3059
+      is_emscripten_ = true;
+      auto start = reinterpret_cast<const uint8_t*>(metadata.data());
+      auto end = reinterpret_cast<const uint8_t*>(metadata.data() + metadata.size());
+      start = decodeVarint(start, end, &emscripten_metadata_major_version_);
+      start = decodeVarint(start, end, &emscripten_metadata_minor_version_);
+      start = decodeVarint(start, end, &emscripten_abi_major_version_);
+      start = decodeVarint(start, end, &emscripten_abi_minor_version_);
+      uint32_t temp;
+      if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 1) {
+        // metadata 0.2 - added: wasm_backend.
+        start = decodeVarint(start, end, &temp);
+      }
       start = decodeVarint(start, end, &temp);
       start = decodeVarint(start, end, &temp);
-      decodeVarint(start, end, &temp);
-      if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 2) {
-        // metadata 0.3 - added: standalone_wasm.
-        start = decodeVarint(start, end, &emscripten_standalone_wasm_);
+      if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 0) {
+        // metadata 0.1 - added: global_base, dynamic_base, dynamictop_ptr and tempdouble_ptr.
+        start = decodeVarint(start, end, &temp);
+        start = decodeVarint(start, end, &temp);
+        start = decodeVarint(start, end, &temp);
+        decodeVarint(start, end, &temp);
+        if (emscripten_metadata_major_version_ > 0 || emscripten_metadata_minor_version_ > 2) {
+          // metadata 0.3 - added: standalone_wasm.
+          start = decodeVarint(start, end, &emscripten_standalone_wasm_);
+        }
       }
     }
+
+    code_ = code;
+    allow_precompiled_ = allow_precompiled;
+  }
+
+  if (started_from_ < Cloneable::InstantiatedModule) {
```

Comment:
!= instead of < as that provides more information in this context and is equivalent.

Reply:
And below

Reply:
Done.